activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -195,9 +196,9 @@ def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, bind
       alias create insert
 
       # Executes the update statement and returns the number of rows affected.
```

Comment:
If I'm following the changes correctly the biggest concern with this PR is going to be the changed signature of `update` and `exec_update` methods. It used to return the number of rows affected but now it suddenly returns a rich result object. Returning the result may be a better design overall but it can't happen without a deprecation. 
Based on the changes we made to `exec_insert` we have the following options:
- Return rich result but only if `returning` argument is non-nil
- new method that always returns rich result which will be used by `_update_record` 
I don't feel completely comfortable of the same method returning different types of values depending on arguments passing but I imagine a new method will raise even more concerns so I think returning result only if `returning` is present is a better option

Reply:
Yes that's definitely a concern and I wasn't sure how the rails team would prefer to approach it.

> Return rich result but only if returning argument is non-nil

This sounds like a pragmatic compromise to me.